jsonalchemy: initial port from Invenio

jirikuncar opened the pull request:
- [x] remove lazy import
- [x] use ~~entry points~~ packages for extensions?

cc @tiborsimko @egabancho

egabancho committed:
JsonAlchemy: initial release

* Removes invenio/core.

* Removes invenio/legacy/bibfield.

Co-authored-by: Adrian-Tudor Panescu <adrian.tudor.panescu@cern.ch>
Co-authored-by: Jiri Kuncar <jiri.kuncar@cern.ch>

NOTE: beware, patch amended by Tibor to use new file names

lnielsen committed:
registry: move to separate package Flask-Registry

* Moves Flask-Registry to a separate external package.

Signed-off-by: Lars Holm Nielsen <lars.holm.nielsen@cern.ch>

NOTE: beware, patch amended by Tibor to use new file names

egabancho committed:
JSONAlchemy: allow `extend` on parser extension

* Allow to extend the content of the parser extensions, like for the
  producer rules.

Signed-off-by: Esteban J. G. Gabancho <esteban.gabancho@cern.ch>

NOTE: beware, patch amended by Tibor to use new file names

egabancho committed:
JSONAlchemy: bug fixes

NOTE: beware, patch amended by Tibor to use new file names

egabancho committed:
JSONAlchemy: bug fixes and tests improvements

Signed-off-by: Esteban J. G. Gabancho <esteban.gabancho@gmail.com>

NOTE: beware, patch amended by Tibor to use new file names

kneczaj committed:
jsonalchemy: exception messages improvements

* Passes detailed parsing error message

Signed-off-by: Kamil Neczaj <kamil.neczaj@cern.ch>

egabancho committed:
jsonalchemy: enhance default value search

* Additionally to setting the default value, the meta-metadata for the
  field is guessed from the field definitions following this rule
  order: `creator`, `derived`, `calculated`. If none of them is found
  every meta-metadad field is set to `UNKNOWN`.

* Convenience property `additional_info` added to SmartJson

Signed-off-by: Esteban J. G. Gabancho <egabancho@gmail.com>

egabancho committed:
records: enhance the API to create empty records

Signed-off-by: Esteban J. G. Gabancho <egabancho@gmail.com>

Commit:
jsonalchemy: JSON-LD addition

* Add utilities for serializing SmartJson objects to JSON-LD. See specification
[1] and the used pyld library [2]. The recommended method for publishing
JSON-LD is via the REST extension.

[1] http://www.w3.org/TR/json-ld/
[2] https://github.com/digitalbazaar/pyld

Signed-off-by: Adrian-Tudor Panescu <adrian.tudor.panescu@cern.ch>

egabancho committed:
jsonalchemy: refactoring fixes

* Fixes issues in jsonalchemy found during testing deposit module.

Co-authored-by: Lars Holm Nielsen <lars.holm.nielsen@cern.ch>
Signed-off-by: Lars Holm Nielsen <lars.holm.nielsen@cern.ch>

jirikuncar committed:
registry: imports from `flask.ext.registry`

Signed-off-by: Jiri Kuncar <jiri.kuncar@cern.ch>

egabancho committed:
records: `reset_cache` added to `api.get_record`

Signed-off-by: Esteban J. G. Gabancho <esteban.gabancho@gmail.com>
Reviewed-by: Jiri Kuncar <jiri.kuncar@cern.ch>

jirikuncar committed:
jsonalchemy: versionable extension

Signed-off-by: Jiri Kuncar <jiri.kuncar@cern.ch>
Co-authored-by: Esteban J. G. Gabancho <esteban.gabancho@cern.ch>

egabancho committed:
jsonalchemy: update `parser.py` for pyparsing 2

* Transform hardcoded decorators into extensions inside the parser
  folder. The new decorators are cataloged as `before`, `on` and `after`
  depending on when they should be evaluated.

* All the decorators that before needed always a tuple now works as
  `@decorator(*args)`.

* Remove `@inherit_from` decorator from field definitions as it is
  replaced by `@extend` and `@override`.

* New decorator extension added `@connect(field_name[, function])` to
  link to fields and propagate changes from one to the other and vice
  versa.

* Remove `include` form field definitions as now registries are used.

* Replace `checker` section by `schema` to perform simple validations,
  like type or length. For more complex check use the checker module.

Signed-off-by: Esteban J. G. Gabancho <esteban.gabancho@gmail.com>

egabancho committed:
jsonalchemy: add `jsonext` as common namespace

Signed-off-by: Esteban J. G. Gabancho <esteban.gabancho@gmail.com>

egabancho committed:
jsonalchemy: update readers and SmartJson

* Update all parsers to work with the new structutre of the field and
  model definitions, meaning using the extensions properly.

* When translating a new blob there is no need any more to go through
  the `readers` registry, instead call `Reader.translate`. See the
  method docs for more info.

* Update SmartJson to work also with new extensions and decorators
  structure.

* Update JSON for marc producer to adapt it to the new extension
  behavior.

* Add first version of sphinx documentation (to be completed).

Signed-off-by: Esteban J. G. Gabancho <esteban.gabancho@gmail.com>

Commit:
jsonalchemy: JSON-LD tweaks

* Adds "inline" synonim for "full" export format to comply with
  specification.

* Fixes __meta_metadata__ cleanup at JSON-LD export time.

Signed-off-by: Adrian-Tudor Panescu <adrian.tudor.panescu@cern.ch>
Reviewed-by: Jiri Kuncar <jiri.kuncar@cern.ch>

jirikuncar committed:
global: six iteritems usage

Signed-off-by: Jiri Kuncar <jiri.kuncar@cern.ch>

jirikuncar committed:
global: exception syntax fix for Python 3

Signed-off-by: Jiri Kuncar <jiri.kuncar@cern.ch>

egabancho committed:
uploader: initial release of insert mode

* NOTE This is an initial release of new uploader using workflows
  and Celery.

Acked-by: Jiri Kuncar <jiri.kuncar@cern.ch>

jirikuncar committed:
jsonalchemy: in memory engine addition

Signed-off-by: Jiri Kuncar <jiri.kuncar@cern.ch>

Commit:
jsonalchemy: import and PEP8 fixes

Co-authored-by: Esteban J. G. Gabancho <esteban.gabancho@gmail.com>
Signed-off-by: Adrian-Tudor Panescu <adrian.tudor.panescu@cern.ch>

jirikuncar committed:
jsonalchemy: fix `six.iteritems` typo

Signed-off-by: Jiri Kuncar <jiri.kuncar@cern.ch>

Commit:
jsonalchemy: UUID validation fix

Signed-off-by: Adrian-Tudor Panescu <adrian.tudor.panescu@cern.ch>

egabancho committed:
jsonalchemy: bug fixes

* Fix record tests when using '001' marc tag.

* Implement `_prepare_blob` method for json_reader to use the model
  field names when using a JSON input.

* Fix model inheritance when overriding fields.

* Fix validation when using models and override fields.

* Enhance `creation_date` field definition.

* PEP8 fixes.

Signed-off-by: Esteban J. G. Gabancho <egabancho@gmail.com>

jirikuncar committed:
jsonalchemy: `uuid` and `objectid` validator fix

Signed-off-by: Jiri Kuncar <jiri.kuncar@cern.ch>

jirikuncar committed:
jsonalchemy: validator test fix

Signed-off-by: Jiri Kuncar <jiri.kuncar@cern.ch>

egabancho committed:
jsonalchemy: model resolver fix

Signed-off-by: Esteban J. G. Gabancho <egabancho@gmail.com>

egabancho committed:
jsonalchemy: extension model fix

Signed-off-by: Esteban J. G. Gabancho <esteban.gabancho@gmail.com>

egabancho committed:
records: aggregation field definitions fix

Signed-off-by: Esteban J. G. Gabancho <egabancho@gmail.com>

jirikuncar committed:
documents: acl extension usage

Signed-off-by: Jiri Kuncar <jiri.kuncar@cern.ch>

jirikuncar committed:
jsonalchemy: memory engine search method addition

Signed-off-by: Jiri Kuncar <jiri.kuncar@cern.ch>

egabancho committed:
jsonalchemy: enhance extension parser behavior

* One extension could be either a importable string pointing to a class
  object or a function call that returns a class object.

* The extensions are now listed as quoted strings, either single quotes
  or double quotes due to some limitations of pyparsing.

Signed-off-by: Esteban J. G. Gabancho <egabancho@gmail.com>
Reviewed-by: Jiri Kuncar <jiri.kuncar@cern.ch>

lnielsen committed:
jsonalchemy: failing test fix

* Fixes failing test cases which was due to side effects introduced
  in 0a7ab631883fe514a28116fbc7fb8b47bd5c1bb4.

* Fixes PEP8 errors.

Signed-off-by: Lars Holm Nielsen <lars.holm.nielsen@cern.ch>

jirikuncar committed:
jsonalchemy: `StorageEngine` metaclass addition

* Adds metaclass that reads `__storagename__` class attribute used as
  prefix for reading the configuration.

Reviewed-by: Lars Holm Nielsen <lars.holm.nielsen@cern.ch>
Signed-off-by: Jiri Kuncar <jiri.kuncar@cern.ch>

jirikuncar committed:
jsonalchemy: Versionable test addition

* Fixes @connect decorator by adding implicit @parse_first
  rule for connected field.

Signed-off-by: Jiri Kuncar <jiri.kuncar@cern.ch>

jirikuncar committed:
jsonalchemy: fixes for versionable extension

Signed-off-by: Jiri Kuncar <jiri.kuncar@cern.ch>

greut committed:
docs: documentation refactoring

* fixing the images' path.

* removing references to non-existing files.

* moving modules into proper files.

Signed-off-by: Yoan Blanc <yoan.blanc@cern.ch>

jirikuncar committed:
jsonalchemy: cache engine addition

Signed-off-by: Jiri Kuncar <jiri.kuncar@cern.ch>

PXke committed:
jsonalchemy: hotfix for optional fields

* Fixes an unexpected behaviour which happen when a field in the
  conversion sheet is not present in the marcxml (e.g. email).  When
  marcxml is transformed to dict everything is fine, but when it is
  converted back to marcxml, it will get into an infinite loop because
  the field is not present in the dict. Hence it will try to eval the
  field and it will end in an infinite import loop.

Signed-off-by: Guillaume Lastecoueres <guillaume.lastecoueres@cern.ch>
Reviewed-by: Jiri Kuncar <jiri.kuncar@cern.ch>

PXke committed:
jsonalchemy: unit tests for module import fix

* Adds a field to fields.cfg, which is a name of module.

* Adds a test that will try to convert a marcxml to json
  and then from json to marcxml. In this test email is
  a field which is not present in the marcxml, and then
  bibfield will try to import it like a module.

Signed-off-by: Guillaume Lastecoueres <guillaume.lastecoueres@cern.ch>
Reviewed-by: Jiri Kuncar <jiri.kuncar@cern.ch>

lnielsen committed:
jsonalchemy: deprecation warning fix

* Fixes a deprecation warning.

Signed-off-by: Lars Holm Nielsen <lars.holm.nielsen@cern.ch>

jalavik committed:
jsonalchemy: create_record error catching

* The marc reader and wrapper modules now catches any errors
  while parsing MARCXML with legacy create_record and returns a
  ReaderException with details.

Signed-off-by: Jan Aage Lavik <jan.age.lavik@cern.ch>

egabancho committed:
jsonalchemy: fix for `__additional_info__` access

Signed-off-by: Esteban J. G. Gabancho <esteban.gabancho@gmail.com>

egabancho committed:
jsonalchemy: default values for subfields

* Fixes jsonalchemy ignoring default values when the main field is
  present but not one of the subfields.

Signed-off-by: Esteban J. G. Gabancho <esteban.gabancho@gmail.com>

greut committed:
docs: fix of sphinx warnings.

* Fixing all the warnings given by sphinx during compilation.

* Fixing the PEP8, PyFlakes and PEP257 warnings of the touched files.

* Adding some intersphinx.

* Reading the theme from docs/requirements.txt.

Signed-off-by: Yoan Blanc <yoan.blanc@cern.ch>

egabancho committed:
jsonalchemy: move to `isinstace(foo, Mapping)`

* Improve type checks to use `collection.Mapping` instead of `dict` as
  this the recommended way and it gives more abstraction.

Signed-off-by: Esteban J. G. Gabancho <esteban.gabancho@gmail.com>

egabancho committed:
jsonalchemy: preserving original tags inside JSON

* When using creator rules that involves regex like `245..` the original
  marc tag is preserved inside the `__meta_metadata__` key for posterior
  needs.

* Enhances `json_for_marc` producer to use the input tag if the
  producer is defined to do so (See `json_for_marc` documentation).
  (addresses #1722)

Signed-off-by: Esteban J. G. Gabancho <esteban.gabancho@gmail.com>

egabancho committed:
records: move legacy methods to the Record object

* Moves `legacy_export_as_marc` and `legacy_create_recstruct` from the
  `SmartJson` class to the record class.

* Enhances test cases for this two methods.

Signed-off-by: Esteban J. G. Gabancho <esteban.gabancho@gmail.com>

egabancho committed:
uploader: field definitions enhancement

Acked-by: Jiri Kuncar <jiri.kuncar@cern.ch>

greut committed:
docs: nit-picky documentation

* Adds nit-picky sphinx-build in travis build.

* Adds placeholder documents for the missing pages.

* Comments out all the badly documented modules.

* NOTE you can test that the documentation is fine by building
  it using: $ sphinx-build -nWE docs docs/_build/html

Signed-off-by: Yoan Blanc <yoan.blanc@cern.ch>

egabancho committed:
jsonalchemy: dirty fix for default values

* FIXME Find a better way to set the default values.

Acked-by: Jiri Kuncar <jiri.kuncar@cern.ch>

egabancho committed:
jsonalchemy: documentation release

jirikuncar committed:
jsonalchemy: documentation and PEP8 fixes

Signed-off-by: Jiri Kuncar <jiri.kuncar@cern.ch>

egabancho committed:
docs: jsonalchemy field definition documentation

jirikuncar committed:
jsonalchemy: support for storage create/drop

* Adds registry for storage engines.

Signed-off-by: Jiri Kuncar <jiri.kuncar@cern.ch>

lnielsen committed:
docs: license inclusion

* Adds license statement to documentation files.

* Fixes wrong license statement in certain docs files.

Signed-off-by: Lars Holm Nielsen <lars.holm.nielsen@cern.ch>

jirikuncar committed:
jsonalchemy: cache engine search fix

Signed-off-by: Jiri Kuncar <jiri.kuncar@cern.ch>

Commit:
jsonalchemy: dumps with specified keywords

* Adds an option to dump only specified keys from dictionary.

Signed-off-by: Leonardo Rossi <leonardo.r@cern.ch>
Reviewed-by: Jiri Kuncar <jiri.kuncar@cern.ch>

Commit:
jsonalchemy: fix SmartJson dumps documentation

* Update dumps documentation to describe the new introduced
  parameter.

Signed-off-by: Leonardo Rossi <leonardo.r@cern.ch>

egabancho committed:
jsonalchemy: validation fixes

Reviewed-by: Jiri Kuncar <jiri.kuncar@cern.ch>
Signed-off-by: Esteban J. G. Gabancho <esteban.gabancho@gmail.com>

Commit:
docs: jsonalchemy grammar and rewording

* Rewords sentences that were incomprehensible or hard to understand
  and splits long chunks of information into bullet points.

* Changes informal pronouns to formal.

* Fixes a triple-quote.

Signed-off-by: Dimitrios Semitsoglou-Tsiapos <dsemitso@cern.ch>

Commit:
authors: base record

* Defines fields and models for author base record.

* Adds new function for jsonalchemy models definitions.

* Includes tests for author base record.

Signed-off-by: Mateusz Susik <mateusz.susik@cern.ch>
Reviewed-by: Jiri Kuncar <jiri.kuncar@cern.ch>

jirikuncar committed:
jsonalchemy: fix problem with reserved names

* Addresses issue when `try_to_eval` function was evaluating missing
  subfields to existing global Python functions.  (closes #2593)

Signed-off-by: Jiri Kuncar <jiri.kuncar@cern.ch>

Commit:
jsonalchemy: print statements removal

Signed-off-by: Mateusz Susik <mateusz.susik@cern.ch>
Reviewed-by: Gilles Louppe <gilles.louppe@cern.ch>
Reviewed-by: Jiri Kuncar <jiri.kuncar@cern.ch>

Commit:
jsonalchemy: function for safe conversion to int

* Adds one unit and one integration test.

Signed-off-by: Mateusz Susik <mateusz.susik@cern.ch>
Reviewed-by: Jiri Kuncar <jiri.kuncar@cern.ch>

jirikuncar committed:
jsonalchemy: @hidden decorator addition

* Adds @hidden decorator for field definition and extends `dumps()`
  method with `filter_hidden=False` argument.  (addresses #2197)

* Improves documentation style according to PEP257.

Signed-off-by: Jiri Kuncar <jiri.kuncar@cern.ch>

jirikuncar committed:
jsonalchemy: metadata separation

Signed-off-by: Jiri Kuncar <jiri.kuncar@cern.ch>

tiborsimko commented:
BTW shall we commit past history from the Invenio repository, as we did for "intbitset"? I can again prepare PR for that.

tiborsimko committed:
jsonalchemy: old history import begins here

* JSONALchemy was being previously developed inside the Invenio digital
  repository software package.  This commit starts migration of its
  recent development history from the `invenio` repository to this
  standalone `jsonalchemy` repository.  This is mostly of historical
  interest, should some changes be required.

* NOTE Before being nicknamed JSONAlchemy, the module started its life
  as a BibField module of Invenio.  This development history is not
  transferred here; please refer to the history of Invenio at
  <http://github.com/inveniosoftware/invenio> if you like to access it.

* NOTE The code is migrated "as is", hence not runnable out of the box.
  Please look for a later commit, headlined "jsonalchemy: import old
  history ends here", which will end the import of the historical code.
  The subsequent commit, entitled "jsonalchemy: initial package
  release", will make the JSONAlchemy code fully runnable as a
  standalone package.

Signed-off-by: Tibor Simko <tibor.simko@cern.ch>

tiborsimko committed:
jsonalchemy: old history import ends here

* This commit completes the import of recent JSONAlchemy development
  history from the Invenio source code repository.  The next commit will
  make `jsonalchemy` into a truly standalone package.  See
  f77ec0b5272e87c3d330aef771cbb836c982beb7 for more explanations.

Signed-off-by: Tibor Simko <tibor.simko@cern.ch>

jirikuncar committed:
general: move to new code structure

egabancho commented:
We should be careful with the producer rules, because we have for example `245..` will take any indicator, but when we produce the output we will produce `245__`.
Simply removing the marc tag and leaving only the subfield should output the same as in the input

jirikuncar commented:
@egabancho please create new issue from https://github.com/inveniosoftware/jsonalchemy/pull/1#issuecomment-73717894
